File: archive/f/f-sharp/BubbleSort.fs
open System


module BubbleSort =
    let private sort (arr: int array) =
        let result = Array.copy arr
        let mutable lastUnsorted = result.Length - 1
        let mutable swapped = true

        while swapped && lastUnsorted > 0 do
            swapped <- false

            for i = 0 to lastUnsorted - 1 do
                if result.[i] > result.[i + 1] then
                    let tmp = result.[i]
                    result.[i] <- result.[i + 1]
                    result.[i + 1] <- tmp
                    swapped <- true

            lastUnsorted <- lastUnsorted - 1

        result

    let run (numbers: int array) =
        numbers |> sort |> Array.map string |> String.concat ", " |> Ok


module Helpers =
    let private (|IntList|_|) (s: string) =
        if String.IsNullOrWhiteSpace(s) then
            None
        else
            let parts = s.Split([| ',' |], StringSplitOptions.RemoveEmptyEntries)
            let len = parts.Length

            if len < 2 then
                None
            else
                let numbers = Array.zeroCreate len
                let mutable ok = true

                for i = 0 to len - 1 do
                    match Int32.TryParse(parts.[i].Trim()) with
                    | true, n -> numbers.[i] <- n
                    | _ -> ok <- false

                if ok then Some numbers else None

    let parseArgs argv =
        match argv with
        | [| IntList numbers |] -> Ok numbers
        | _ -> Error "Usage: please provide a list of at least two integers to sort in the format \"1, 2, 3, 4, 5\""

    let handleResult =
        function
        | Ok result ->
            printfn "%s" result
            0
        | Error msg ->
            eprintfn "%s" msg
            1

[<EntryPoint>]
let main argv =
    argv |> Helpers.parseArgs |> Result.bind BubbleSort.run |> Helpers.handleResult

File: archive/c/c-sharp/BubbleSort.cs
using System.Collections.Generic;

if (args is not [var input])
    return ExitWithUsage();

if (!TryParseList(input.AsSpan(), out var numbers))
    return ExitWithUsage();

BubbleSort(numbers);

Console.WriteLine(string.Join(", ", numbers));
return 0;

static bool TryParseList(ReadOnlySpan<char> view, out List<int> numbers)
{
    numbers = new List<int>();

    while (!view.IsEmpty)
    {
        int comma = view.IndexOf(',');

        ReadOnlySpan<char> segment = comma < 0 ? view : view[..comma];

        segment = segment.Trim();

        if (segment.IsEmpty || !int.TryParse(segment, out int val))
            return false;

        numbers.Add(val);

        if (comma < 0)
            break;

        view = view[(comma + 1)..];
    }

    return numbers.Count >= 2;
}

static void BubbleSort(List<int> xs)
{
    int n = xs.Count;

    for (int i = 0; i < n - 1; i++)
    {
        bool swapped = false;

        for (int j = 0; j < n - i - 1; j++)
        {
            if (xs[j] > xs[j + 1])
            {
                (xs[j], xs[j + 1]) = (xs[j + 1], xs[j]);
                swapped = true;
            }
        }

        if (!swapped)
            break;
    }
}

static int ExitWithUsage()
{
    Console.WriteLine(
        "Usage: please provide a list of at least two integers to sort in the format \"1, 2, 3, 4, 5\""
    );
    return 1;
}
